how-to/get-started: describe how to connect to a testnet

q9f opened the pull request:
## Pull Request Description
This is the first draft of how to choose and connect to a testnet. I haven't spell-checked it yet.

## Fixed Issue(s)
Fixes  #10

q9f commented on mkdocs.yml:
```diff
@@ -95,8 +96,8 @@ markdown_extensions:
-  - pymdownx.superfences:
-      legacy_tab_classes: true
+  ### - pymdownx.superfences:
+  ###    legacy_tab_classes: true
```
I have no clue how to make `mkdocs serve` work with this option.

NicolasMassart replied:
Did you try to run it in a Python virtual environment and run `pip install -r common/build_tools/requirements.txt` ?

q9f replied:
yes

```
~/.opt/byz-f/doc.teku q9-teku-testnet
❯ pip install -r common/build_tools/requirements.txt
Defaulting to user installation because normal site-packages is not writeable
Requirement already satisfied: mkdocs==1.1 in /home/user/.local/lib/python3.8/site-packages (from -r common/build_tools/requirements.txt (line 1)) (1.1)
Requirement already satisfied: mkdocs-material==4.6.3 in /home/user/.local/lib/python3.8/site-packages (from -r common/build_tools/requirements.txt (line 2)) (4.6.3)
Requirement already satisfied: Markdown==3.2.1 in /home/user/.local/lib/python3.8/site-packages (from -r common/build_tools/requirements.txt (line 3)) (3.2.1)
Requirement already satisfied: markdown-fenced-code-tabs==1.0.3 in /home/user/.local/lib/python3.8/site-packages (from -r common/build_tools/requirements.txt (line 4)) (1.0.3)
Requirement already satisfied: markdown-include==0.5.1 in /home/user/.local/lib/python3.8/site-packages (from -r common/build_tools/requirements.txt (line 5)) (0.5.1)
Requirement already satisfied: MarkupSafe==1.1.1 in /usr/lib/python3.8/site-packages (from -r common/build_tools/requirements.txt (line 6)) (1.1.1)
Requirement already satisfied: mkdocs-markdownextradata-plugin==0.1.3 in /home/user/.local/lib/python3.8/site-packages (from -r common/build_tools/requirements.txt (line 7)) (0.1.3)
Requirement already satisfied: mkdocs-exclude==1.0.2 in /home/user/.local/lib/python3.8/site-packages (from -r common/build_tools/requirements.txt (line 8)) (1.0.2)
Requirement already satisfied: mkdocs-redirects==1.0.0 in /home/user/.local/lib/python3.8/site-packages (from -r common/build_tools/requirements.txt (line 9)) (1.0.0)
Requirement already satisfied: mkdocs-git-revision-date-localized-plugin==0.5.0 in /home/user/.local/lib/python3.8/site-packages (from -r common/build_tools/requirements.txt (line 10)) (0.5.0)
Requirement already satisfied: mkdocs-minify-plugin==0.3.0 in /home/user/.local/lib/python3.8/site-packages (from -r common/build_tools/requirements.txt (line 11)) (0.3.0)
Requirement already satisfied: livereload>=2.5.1 in /usr/lib/python3.8/site-packages (from mkdocs==1.1->-r common/build_tools/requirements.txt (line 1)) (2.6.1)
Requirement already satisfied: click>=3.3 in /usr/lib/python3.8/site-packages (from mkdocs==1.1->-r common/build_tools/requirements.txt (line 1)) (7.1.1)
Requirement already satisfied: Jinja2>=2.10.1 in /usr/lib/python3.8/site-packages (from mkdocs==1.1->-r common/build_tools/requirements.txt (line 1)) (2.11.2)
Requirement already satisfied: lunr[languages]==0.5.6 in /home/user/.local/lib/python3.8/site-packages (from mkdocs==1.1->-r common/build_tools/requirements.txt (line 1)) (0.5.6)
Requirement already satisfied: tornado>=5.0 in /usr/lib/python3.8/site-packages (from mkdocs==1.1->-r common/build_tools/requirements.txt (line 1)) (6.0.4)
Requirement already satisfied: PyYAML>=3.10 in /usr/lib/python3.8/site-packages (from mkdocs==1.1->-r common/build_tools/requirements.txt (line 1)) (5.3.1)
Requirement already satisfied: Pygments>=2.4 in /usr/lib/python3.8/site-packages (from mkdocs-material==4.6.3->-r common/build_tools/requirements.txt (line 2)) (2.6.1)
Requirement already satisfied: pymdown-extensions>=6.3 in /usr/lib/python3.8/site-packages (from mkdocs-material==4.6.3->-r common/build_tools/requirements.txt (line 2)) (6.3)
Requirement already satisfied: setuptools>=36 in /usr/lib/python3.8/site-packages (from Markdown==3.2.1->-r common/build_tools/requirements.txt (line 3)) (46.1.3)
Requirement already satisfied: htmlmin>=0.1.12 in /usr/lib/python3.8/site-packages (from markdown-fenced-code-tabs==1.0.3->-r common/build_tools/requirements.txt (line 4)) (0.1.12)
Requirement already satisfied: babel>=2.7.0 in /home/user/.local/lib/python3.8/site-packages (from mkdocs-git-revision-date-localized-plugin==0.5.0->-r common/build_tools/requirements.txt (line 10)) (2.8.0)
Requirement already satisfied: GitPython in /home/user/.local/lib/python3.8/site-packages (from mkdocs-git-revision-date-localized-plugin==0.5.0->-r common/build_tools/requirements.txt (line 10)) (3.1.2)
Requirement already satisfied: jsmin>=2.2.2 in /usr/lib/python3.8/site-packages (from mkdocs-minify-plugin==0.3.0->-r common/build_tools/requirements.txt (line 11)) (2.2.2)
Requirement already satisfied: six in /usr/lib/python3.8/site-packages (from livereload>=2.5.1->mkdocs==1.1->-r common/build_tools/requirements.txt (line 1)) (1.14.0)
Requirement already satisfied: future>=0.16.0 in /home/user/.local/lib/python3.8/site-packages (from lunr[languages]==0.5.6->mkdocs==1.1->-r common/build_tools/requirements.txt (line 1)) (0.18.2)
Requirement already satisfied: nltk>=3.2.5; extra == "languages" in /home/user/.local/lib/python3.8/site-packages (from lunr[languages]==0.5.6->mkdocs==1.1->-r common/build_tools/requirements.txt (line 1)) (3.5)
Requirement already satisfied: pytz>=2015.7 in /home/user/.local/lib/python3.8/site-packages (from babel>=2.7.0->mkdocs-git-revision-date-localized-plugin==0.5.0->-r common/build_tools/requirements.txt (line 10)) (2020.1)
Requirement already satisfied: gitdb<5,>=4.0.1 in /home/user/.local/lib/python3.8/site-packages (from GitPython->mkdocs-git-revision-date-localized-plugin==0.5.0->-r common/build_tools/requirements.txt (line 10)) (4.0.5)
Requirement already satisfied: joblib in /home/user/.local/lib/python3.8/site-packages (from nltk>=3.2.5; extra == "languages"->lunr[languages]==0.5.6->mkdocs==1.1->-r common/build_tools/requirements.txt (line 1)) (0.14.1)
Requirement already satisfied: regex in /home/user/.local/lib/python3.8/site-packages (from nltk>=3.2.5; extra == "languages"->lunr[languages]==0.5.6->mkdocs==1.1->-r common/build_tools/requirements.txt (line 1)) (2020.4.4)
Requirement already satisfied: tqdm in /home/user/.local/lib/python3.8/site-packages (from nltk>=3.2.5; extra == "languages"->lunr[languages]==0.5.6->mkdocs==1.1->-r common/build_tools/requirements.txt (line 1)) (4.46.0)
Requirement already satisfied: smmap<4,>=3.0.1 in /home/user/.local/lib/python3.8/site-packages (from gitdb<5,>=4.0.1->GitPython->mkdocs-git-revision-date-localized-plugin==0.5.0->-r common/build_tools/requirements.txt (line 10)) (3.0.4)

~/.opt/byz-f/doc.teku q9-teku-testnet*
❯ mkdocs serve                                      
INFO    -  Building documentation... 
ERROR   -  Config value: 'markdown_extensions'. Error: legacy_tab_classes 

Aborted with 1 Configuration Errors!
```

NicolasMassart commented on docs/HowTo/Get-Started/Connect-To-Testnet.md:
```diff
@@ -0,0 +1,184 @@
+
+**Prerequisites**:
+
+* [Install Teku](/HowTo/Get-Started/Install-Binaries) Ethereum 2.0 client or [Build Teku from Source](/HowTo/Get-Started/Build-From-Source).
```
links have to point to the relative .md source files, not the URLs generated by MkDocs.
Here it should be `/HowTo/Get-Started/Install-Binaries.md` and `/HowTo/Get-Started/Build-From-Source.md`

NicolasMassart commented on docs/HowTo/Get-Started/Connect-To-Testnet.md:
````diff
@@ -0,0 +1,184 @@
+
+Teku allows to select preconfigured networks with the `--network` flag:
+
+```bash
````
Could you try to use the same CLI syntax layout than in https://github.com/PegaSysEng/doc.teku/blob/master/docs/Reference/CLI/CLI-Syntax.md ?

It may be something like:

```
    ```bash tab="Syntax"
    -n, --network=<NETWORK>
    ```
    
    ```bash tab="Command Line"
    --network=mainnet
    ```
    
    ```bash tab="Environment Variable"
    TEKU_network=mainnet
    ```
    
    Represents which network to use.
    The default is `minimal`.
```

q9f replied:
Done that but it completely bricks the rendered version.

![Screenshot_2020-05-07 Connect to a Testnet - Teku](https://user-images.githubusercontent.com/58883403/81289402-a7881a00-9066-11ea-9c50-8a8c80284359.png)

NicolasMassart replied:
Hum yes if the superfences is not activated... Let me try to build your PR locally. I also sent you a ping on Gitter if you have some time to discuss this.

NicolasMassart replied:
Tested locally and it works well for me so I think it's just a small conflict in Python packaging, we already saw that and it's not very hard to fix but requires some patience...
I can help you with that. The thing for me to take away from this is that I have to document the doc building process and doc dev environment a bit more. It's not the same product, but I'm not going to lie, it's written by me too, so if you would like more details on the doc building, you can have a look at https://wiki.hyperledger.org/display/BESU/MkDocs+And+Markdown+Guide#MkDocsAndMarkdownGuide-PreviewTheDocumentation
Except that Besu doesn't use the common Git submodule, it's mostly the same thing, I try to be consistant on the tooling. I also sent you a message on Gitter if you want to chat more interactively.

NicolasMassart commented on docs/HowTo/Get-Started/Connect-To-Testnet.md:
```diff
@@ -0,0 +1,184 @@
+
+!!! tip
+
+    If you don't know which network to pick, choose `--network=schlesi`.
```
could you explain why?

NicolasMassart commented on docs/HowTo/Get-Started/Connect-To-Testnet.md:
```diff
@@ -0,0 +1,184 @@
+
+!!! note
+
+    Note, that building from `master` might come with certain instability and features may break without prior warning.
```
You may not need to repeat "Note," here as it's already the title of the box.

NicolasMassart commented on docs/HowTo/Get-Started/Connect-To-Testnet.md:
```diff
@@ -0,0 +1,184 @@
+
+    Note, that building from `master` might come with certain instability and features may break without prior warning.
+
+Simply pass the `--network=schlesi` command-line flag to Teku.
```
Remove the world "Simply" according to our writing guidelines please. Never consider something simple because it may be for you but not for the reader.

NicolasMassart commented on docs/HowTo/Get-Started/Connect-To-Testnet.md:
```diff
@@ -0,0 +1,184 @@
+
+## Connect to other Testnets
+
+Under certain circumstances it's possible to connect to other public testnets.
```
What do you mean by "Under certain circumstances" ?

NicolasMassart commented on docs/HowTo/Get-Started/Connect-To-Testnet.md:
```diff
@@ -0,0 +1,184 @@
+
+**Network configuration**:
+
+The network configuration can be either of the presets `minimal` and `mainnet` or a `yaml` file. The custome configuration file can be passed as local file or remote config fetched via HTTP, e.g.:
```
use "for example" instead of "e.g." (used multiple times in the file, not only on this line)

NicolasMassart commented on docs/HowTo/Get-Started/Connect-To-Testnet.md:
````diff
@@ -0,0 +1,184 @@
+    teku --network="https://github.com/goerli/schlesi/raw/master/teku/chain.yaml"
+    ```
+
+**Deposit Contract**:
````
Fix title case to sentence case. Only the first word and proper nouns should start with a capital letter.
You also have to do it in other titles, as for example, "testnet" is not a proper noun or trademark and should be lowercased.

NicolasMassart commented on docs/HowTo/Get-Started/Connect-To-Testnet.md:
```diff
@@ -0,0 +1,184 @@
+
+Note, that reading the deposit contract requires an enabled ETH1 connection, e.g.:
+
+`--eth1-enabled --eth1-endpoint="localhost:8545"`
```
only one example line may be ok. I would prefer you to keep only the second full one.

NicolasMassart commented on docs/HowTo/Get-Started/Connect-To-Testnet.md:
```diff
@@ -0,0 +1,184 @@
+
+**Genesis State**:
+
+In case the genesis state is already known and available, instead of passing an deposit contract, the initial state can be provided:
```
Write "passing **a** deposit contract" instead of "passing **an** deposit contract"

NicolasMassart commented on docs/HowTo/Get-Started/Connect-To-Testnet.md:
```diff
@@ -0,0 +1,184 @@
+* `--p2p-discovery-bootnodes="enr:-LK4QFO0gKFieMiNrUystSk5Xt7DmIgusloLudv-gH8Krjw9SsUDZRk---H-3hwvL9rMfsMcZwU6L5ezK2d1_dG0UgECh2F0dG5ldHOIAAAAAAAAAACEZXRoMpCZJe_WAAAAAP__________gmlkgnY0gmlwhDMPd52Jc2VjcDI1NmsxoQPNb3TG-iN0aGTagN4peO0SEkWKklJOvloWL0He8pnB_4N0Y3CCJRyDdWRwgiUc"`
+* `--p2p-static-peers="/ip4/51.15.119.157/tcp/9000/p2p/16Uiu2HAkvLCWwVEfF365ZWXB6siDL1mUpcd1XQ1nSXAHmvM5W7wn"`
+
+If the number of command-line arguments becomes confusing, it's recommended to [Use a Configuration File](HowTo/Configure/Use-Configuration-File.md).
```
lowercase "Use a Configuration File"

NicolasMassart replied:
Also the link is wrong (see CI linkchecker report on this PR). It should be `/HowTo/Configure/Use-Configuration-File.md`, starting with a `/` as otherwise it's relative to the current page position and then points to `/HowTo/Get-Started/HowTo/Configure/Use-Configuration-File.md` which doesn't exist.

q9f commented:
Addressed all comments and ran a spell-checker. I think it's good to be reviewed now. 

I still have a couple of issues with `mkdocs serve`:
* does not like the superfences config
* relative links are 404 on local build
* the tabbed syntax thingy is not rendering locally

NicolasMassart commented:
I tested locally and `mkdocs serve` and even strict building passes. So I think it's just a small conflict in Python packaging, we already saw that and it's not very hard to fix but requires some patience and I can take time to help you with that.

if you already have [Circle CI CLI](https://circleci.com/docs/2.0/local-cli/) installed locally, you can try running the build using the CI helper script in `common/build_tools/scripts/test_build.sh` and you will be able to notice it works as it uses containers and so it's going to pass by issues on your local machine and you should see a gree `Success!` in the end.

The thing for me to take away from this is that I have to document the doc building process and doc dev environment a bit more.
Besu is not the same product (but I'm not going to lie,doc tools doc is written by me too) so if you would like more details on the doc building, you can have a look at https://wiki.hyperledger.org/display/BESU/MkDocs+And+Markdown+Guide#MkDocsAndMarkdownGuide-PreviewTheDocumentation
Except that Besu doesn't use the common Git submodule, it's mostly the same thing, I try to be consistent on the tooling.

I also sent you a message on Gitter if you want to chat more interactively.

bgravenorst commented:
Thanks for you contribution to the Teku docs! We'll approve this pull request and implement any required changes ourselves, rather than go back and forth with editing suggestions.
